xwiki-commons-core/xwiki-commons-observation/xwiki-commons-observation-local/src/main/java/org/xwiki/observation/internal/DefaultObservationManager.java

```diff
@@ -131,7 +131,9 @@ void removeEvent(Event event)
             initializeListeners();
         }
 
-        return this.listenersByEvent;
+        synchronized (this.listenersByEvent) {
```

Comment:
This is really not great for performance since it adds locking on read (notify). If you really need something like this then you should use a ReadWriteLock instead.

Also this lock does not have any impact on removeListener which uses getListenersByName(), are you sure it's enough ?

Reply:
> This is really not great for performance since it adds locking on read (notify)

actually that's why I preferred the solution to put the synchronize in add/remove listener to avoid a bit perf issues.

> Also this lock does not have any impact on removeListener which uses getListenersByName(), are you sure it's enough ?

I'm sure I did not reproduce the bug with it. But AFAIU the bug is linked with adding a listener, not removing, so it makes sense.

Reply:
> I'm sure I did not reproduce the bug with it. But AFAIU the bug is linked with adding a listener, not removing, so it makes sense.

In previous version you did not had any locking on notify side. That would mean the issue is concurrency issue between several threads doing addListener only ?

Reply:
> That would mean the issue is concurrency issue between several threads doing addListener only ?

I think so, yes. That's why it appears very rarely IMO

Reply:
OK but it also this the synchronoize at this level is not needed at all and having one at addListener level should be enough.

Reply:
not sure to understand: you mean I should have only one synchronize in addListener, and nothing elsewhere, right?